zerver/actions/message_edit.py

```diff
@@ -147,18 +147,27 @@ def maybe_send_resolve_topic_notifications(
     old_topic_name: str,
     new_topic_name: str,
     changed_messages: QuerySet[Message],
+    pre_truncation_topic_name: str,
 ) -> Tuple[Optional[int], bool]:
     """Returns resolved_topic_message_id if resolve topic notifications were in fact sent."""
     # Note that topics will have already been stripped in check_update_message.
     #
     # This logic is designed to treat removing a weird "✔ ✔✔ "
     # prefix as unresolving the topic.
-    topic_resolved: bool = new_topic_name.startswith(
-        RESOLVED_TOPIC_PREFIX
-    ) and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
-    topic_unresolved: bool = old_topic_name.startswith(
-        RESOLVED_TOPIC_PREFIX
-    ) and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+    resolved_prefix_len = len(RESOLVED_TOPIC_PREFIX)
+    topic_resolved: bool = (
+        new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+        and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+        and (
+            new_topic_name[resolved_prefix_len:] == old_topic_name
+            or pre_truncation_topic_name != new_topic_name
```

Comment:
I assume you meant to say `pre_truncation_topic_name == new_topic_name`, as otherwise, `topic_resolved` will almost always be set to false when the topic was actually resolved.

 Unfortunately, this implementation does not work. The issue with truncation is that the topic name changes to accommodate the "resolved prefix" when a topic is resolved. This means that the comparison
`new_topic_name[resolved_prefix_len:] == old_topic_name` will fail if the topic was truncated, even though the topic was not moved. Therefore, we need to set the boolean to true if the topic was resolved and truncated.

For a better understanding of the problem watch the video below:
https://github.com/zulip/zulip/assets/160421883/692dfe73-5f21-4dcc-8d74-48c173bc92ef

Reply:
Sorry, I meant to say:

`pre_truncation_topic_name.lstrip(RESOLVED_TOPIC_PREFIX) == old_topic_name`

We can name `pre_truncation_topic_name` to `pre_truncation_new_topic_name` for clarity. -- It is the new topic name without truncation.

Read `pre_truncation_topic_name` as `pre_truncation_new_topic_name` in this screenshot:

![Screenshot from 2024-06-27 08-38-19](https://github.com/zulip/zulip/assets/56781761/7320ee21-698c-47d0-9d3c-2b8372ceaed7)

Reply:
So, a topic is said to be resolved when the `new_topic_name` starts with `RESOLVED_TOPIC_PREFIX`, `old_topic_name` doesn't have `RESOLVED_TOPIC_PREFIX` and the actual name part without that `RESOLVED_TOPIC_PREFIX` symbol is the same.

Reply:
It's past my bedtime where I live :) I'll look into this and the rest when I wake up.